src/solver.py

```diff
@@ -1,21 +1,48 @@
-import argparse
-from model import Grid
+from model.grid import Grid
+from rules import solve_single_candidates
 
-parser = argparse.ArgumentParser(
-    description="A Sudoku solver which applies solving rules just as a human would."
-)
-parser.add_argument("input", help="The sudoku puzzle to solve.")
 
-args = parser.parse_args()
+class Solver:
+    def __init__(self, grid: Grid):
+        """Construct a Solver instance with the given Sudoku grid.
 
+        Args:
+            grid (Grid): The Sudoku grid to be solved.
+        """
+        self.grid = grid
 
-def run():
-    with open(args.input, "r") as f:
-        lines = f.readlines()
-        grid = Grid.from_rows_notation(lines)
+    def solve(self):
+        """Solve the Sudoku puzzle using a cycle of rules until no more rules can be applied."""
+        while True:
+            applied = False
 
-    grid.display()
+            # Apply rules
+            applied |= solve_single_candidates(self.grid)
```

Comment:
Every rule will try to apply to the Grid and will change the `applied` variable to True if any one of them manages to solve at least one cell.  While `applied` remains `True` after applying all rules, we can run through the rules another time.